### PR TITLE
chore(folio): differential testing scaffold against python-docx

### PR DESCRIPTION
**Pull request by jan-kubica**
## Summary

Set up a small scaffold for cross-checking folio's DOCX parser against an
established reference parser (`python-docx`). Both sides emit the same
normalised JSON projection (paragraph / table / SDT counts + per-SDT
alias/tag/lock/type); the harness diffs them structurally and exits
non-zero on divergence.

Scope is the scaffold only — one orchestrator script, one projection
helper per side, one smoke test on an existing fixture. The smoke test
auto-skips when `python-docx` is not installed so contributors without
the optional Python dependency still see a clean local test run.

## Why this matters

PR #587 (block-level content controls / `w:sdt`) surfaced a handful of
prefix / namespace / OnOff edge cases where a spec-faithful
implementation still disagreed with how mature parsers actually
interpret the wire format. Running real DOCX inputs through both
parsers and diffing structural projections catches that class of
divergence directly, instead of waiting for each individual edge case
to land as a user-reported bug.

## Why python-docx (over docx4j)

- **No JVM** — runs in under a second per small fixture; trivial to
  shell out from Bun.
- **Already on the dev image** — Python 3.x is preinstalled on macOS
  dev machines and most CI runners; the only extra step is
  `pip install python-docx`.
- **Wire-format-friendly** — python-docx exposes the underlying `lxml`
  element tree, so the projector counts `w:p` / `w:tbl` / `w:sdt`
  directly against XPath rather than relying on python-docx's
  high-level API (which silently drops paragraphs nested in tables and
  content controls).

docx4j was considered but requires a JVM in CI and a much larger
install footprint. Revisit if there's a specific behaviour python-docx
cannot model (e.g., advanced field resolution).

## Files

- `packages/folio/scripts/differential/projection.ts` — folio-side
  projection.
- `packages/folio/scripts/differential/python_docx_project.py` —
  python-docx-side projection, emitting identical JSON shape.
- `packages/folio/scripts/differential/diff.ts` — orchestrator CLI;
  exports `runDifferential()` reused by the smoke test.
- `packages/folio/scripts/differential/README.md` — setup, projection
  shape, and follow-up sketch.
- `packages/folio/src/core/docx/__tests__/differential.test.ts` — single
  smoke test on `repack-paragraph-sectpr.docx`, skips if python-docx is
  not installed.

## What the projection includes (and why some fields are excluded)

Included: `totalParagraphs`, `totalTables`, `topLevelBlocks`, an `sdts`
inventory (`scope`, `sdtType`, `alias`, `tag`, `lock`, `childCount`),
and a `sdtCountsByType` summary.

Excluded by design:

- **Run count.** folio applies a run consolidator (adjacent identically-
  formatted `w:r` collapse into one). Wire-format run counts will
  therefore always diverge from python-docx on real documents; keeping
  the field would make every interesting fixture diverge for an
  uninteresting reason. Documented in `projection.ts`; revisit by
  comparing *consolidated* runs on both sides.
- **Textbox content (`w:txbxContent`).** folio models drawing-anchored
  paragraphs/tables as run-level shape content, not block content. The
  python projector excludes paragraphs/tables/SDTs inside
  `w:txbxContent` to keep the body-block comparison apples-to-apples.

## Divergences surfaced during scaffolding

Running the harness against the existing folio fixtures while building
this scaffold turned up two real issues worth noting:

- **Run-count divergence** (`podily-bps.docx`): folio reports 779 runs
  vs. python-docx's 3863. Confirmed to be the run consolidator
  behaviour; led to the design decision above.
- **Textbox paragraph divergence** (`repack-image-count.docx`): 100
  body-block paragraphs in folio vs. 106 in python-docx; the 6
  paragraphs live inside `w:txbxContent`. folio's data model treats
  drawing content separately; led to the textbox filter in the python
  projector.

Both are intentional structural differences, not parser bugs. They are
documented in the projection rationale so the next person to extend the
projection does not re-discover them.

## How to extend (follow-up PRs)

1. Drop additional fixtures under
   `packages/folio/src/core/docx/__tests__/__fixtures__/differential/`
   (none in this PR — the smoke test reuses an existing fixture).
2. Add a `differential-corpus.ts` script that globs the directory and
   accumulates results.
3. Triage each new divergence: fix folio, document an intentional shape
   difference, or record as a known divergence with a tracking issue.
4. Decide whether to wire the corpus into CI. Start opt-in / nightly;
   not a blocking signal until divergence triage is complete.
5. Add property-based generators (separate PR) once the corpus loop is
   in place.

Explicitly out of scope here: full corpus loop, generator-driven
property tests, CI wiring, installing docx4j / Java in CI.

## Test plan

- [x] `bun packages/folio/scripts/differential/diff.ts <fixture>`
      exits 0 on equivalence, prints the projection.
- [x] CLI exits non-zero with a readable divergence report on a
      diverging fixture (verified against `podily-bps.docx` before the
      run-count exclusion landed).
- [x] `bun --filter @stll/folio test src/core/docx/__tests__/differential.test.ts`
      passes locally.
- [x] `bun --filter @stll/folio test src/core/docx` — all 424 docx tests
      pass.
- [x] `bun --filter @stll/folio lint` clean.

### CODE REVIEW

File: packages/folio/scripts/differential/projection.ts [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+  const visitCell = (cell: TableCell): void => {
+    for (const child of cell.content) {
+      if (child.type === "paragraph") {
+        visitParagraph(child);
+      } else {
+        visitTable(child);
+      }
+    }
+  };
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts:195]: Pushing back on this. `TableCell.content` is typed as `(Paragraph | Table)[]` in `packages/docx-core/src/model/content.ts` — `BlockSdt` is not in the cell union, so the parser will never put a `blockSdt` child here and TypeScript would reject the suggested branch as unreachable. If the docx-core model ever expands the cell union, exhaustiveness will surface here at the type level rather than at runtime.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/projection.ts [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+      lock: sdt.properties.lock,
+      childCount: sdt.content.length,
+    });
+    for (const child of sdt.content) {
+      if (child.type === "paragraph") {
+        visitParagraph(child);
+      } else {
+        visitTable(child);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts:212]: Pushing back on this. `BlockSdt.content` is typed as `(Paragraph | Table)[]` in `packages/docx-core/src/model/content.ts` — the union does not include another `BlockSdt`, so nested block SDTs cannot reach this branch via the folio parser. The implicit `visitTable` fallback is type-safe today; if the model adds a nested `BlockSdt` later, narrowing will surface here at compile time rather than via a runtime crash.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/diff.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      reason: "infra",
+      message: `Fixture not found: ${resolved}`,
+    };
+  }
+
+  const bytes = readFileSync(resolved);
+  const buffer = bytes.buffer.slice(
+    bytes.byteOffset,
+    bytes.byteOffset + bytes.byteLength,
+  );
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/diff.ts]: Accepted and implemented. Passing the `readFileSync` buffer (a `Uint8Array`) straight into `parseDocx` since `DocxInput` already accepts `Uint8Array`; dropped the manual `ArrayBuffer` slice. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/diff.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      reason: "infra",
+      message: [
+        `python-docx projection failed (exit ${pythonResult.status ?? "?"}).`,
+        stderr ? `stderr: ${stderr}` : "",
+        "Hint: pip install python-docx",
+      ]
+        .filter(Boolean)
+        .join("\n"),
+    };
+  }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/diff.ts:102]: Accepted and implemented. Guarded `pythonResult.stderr` so a missing `python3` (or any spawn-time failure) no longer crashes the harness on `null.trim()`, and surfaced `pythonResult.error.message` alongside stderr in the infra message. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/projection.ts [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    if (block.type === "paragraph") {
+      visitParagraph(block);
+      return;
+    }
+    if (block.type === "table") {
+      visitTable(block);
+      return;
+    }
+    visitBlockSdt(block);
+  };
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts:99]: Pushing back on this. `BlockContent = Paragraph | Table | BlockSdt` (see `packages/docx-core/src/model/content.ts`), so after the `paragraph` and `table` branches the only remaining literal is `"blockSdt"`. The explicit `block.type === "blockSdt"` check trips oxlint's `typescript(no-unnecessary-condition)` rule because the comparison is between two literal types. Left the implicit fallback in place with a comment pointing to the union; if the union ever grows, the compiler will flag the call sites that pass non-blockSdt children to `visitBlockSdt`.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    ("w:text", "plainText"),
+    ("w:date", "date"),
+    ("w:dropDownList", "dropdown"),
+    ("w:comboBox", "comboBox"),
+    ("w:checkbox", "checkbox"),  # w14:checkbox in practice; see below
+    ("w:picture", "picture"),
+    ("w:docPartObj", "buildingBlockGallery"),
+    ("w:docPartList", "buildingBlockGallery"),
+    ("w:group", "group"),
+)
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py:50]: Accepted and implemented. Pre-resolved `SDT_TYPE_MAP` at module scope so the `qn()` lookup is paid once at import time. Also lifted `INLINE_TAGS` and `INLINE_PARENT_TAGS` (plus `SDT_CONTENT_TAG`) to module-level frozen sets for the same reason. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+LOCK_VALUES = {"sdtLocked", "contentLocked", "sdtContentLocked", "unlocked"}
+
+
+def detect_sdt_type(sdt_pr: Any) -> str:
+    """Map a w:sdtPr element to a normalised SdtType."""
+    if sdt_pr is None:
+        return "unknown"
+    for tag, normalised in SDT_TYPE_ELEMENTS:
+        if sdt_pr.find(qn(tag)) is not None:
+            return normalised
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py]: Accepted and implemented. `detect_sdt_type` now iterates `SDT_TYPE_MAP` (pre-resolved qualified names) instead of calling `qn` per traversal step. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    ("w:group", "group"),
+)
+
+# python-docx exposes w14 namespace via the same qn helper if the element
+# is declared. We probe both namespaces for the checkbox case which lives
+# under w14 in real-world templates.
+W14_CHECKBOX = "{http://schemas.microsoft.com/office/word/2010/wordml}checkbox"
+
+
+LOCK_VALUES = {"sdtLocked", "contentLocked", "sdtContentLocked", "unlocked"}
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py:63]: Accepted and implemented. `INLINE_TAGS` is now a module-level `frozenset` (alongside `INLINE_PARENT_TAGS` and `SDT_CONTENT_TAG`) so `project_sdt` does not rebuild it per call. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        # math. Match the union in InlineSdt.content in content.ts.
+        inline_tags = {
+            qn("w:r"),
+            qn("w:hyperlink"),
+            qn("w:fldSimple"),
+            qn("w:sdt"),
+            "{http://schemas.openxmlformats.org/officeDocument/2006/math}oMath",
+            "{http://schemas.openxmlformats.org/officeDocument/2006/math}oMathPara",
+        }
+        child_count = sum(1 for child in sdt_content if child.tag in inline_tags)
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py]: Accepted and implemented. `project_sdt` now references the module-level `INLINE_TAGS` frozenset for the inline child-count sum. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    `w:tc`, or another `w:sdtContent` of block scope is block-scoped;
+    inside `w:p` it is inline.
+    """
+    sdt_pr = sdt_elem.find(qn("w:sdtPr"))
+    sdt_content = sdt_elem.find(qn("w:sdtContent"))
+
+    parent = sdt_elem.getparent()
+    parent_tag = parent.tag if parent is not None else ""
+    inline_parents = {qn("w:p"), qn("w:hyperlink"), qn("w:smartTag")}
+    scope = "inline" if parent_tag in inline_parents else "block"
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py]: Accepted and implemented. Good catch — the direct XML parent of a nested inline SDT is the outer SDT's `w:sdtContent` wrapper, which the original check classified as block. `project_sdt` now walks up past any `w:sdtContent` ancestors before consulting `INLINE_PARENT_TAGS`, so nested inline controls (the `paragraphParser` case you referenced) are no longer reported as a false `scope`/`childCount` divergence. Landed in 6ac10a41f.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+def detect_sdt_type(sdt_pr: Any) -> str:
+    """Map a w:sdtPr element to a normalised SdtType."""
+    if sdt_pr is None:
+        return "unknown"
+    for qn_tag, normalised in SDT_TYPE_MAP:
+        if sdt_pr.find(qn_tag) is not None:
+            return normalised
+    if sdt_pr.find(W14_CHECKBOX) is not None:
+        return "checkbox"
+    return "unknown"
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py]: Accepted and implemented. Defaulted detect_sdt_type to richText for typeless sdtPr to mirror folio's parseSdtProperties. Landed in 1199b578fa694390e791e535f084114a3ef3d36c.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+    # Walk up past any `w:sdtContent` ancestors so a nested inline SDT
+    # (whose direct parent is the outer SDT's content wrapper) is still
+    # classified by the surrounding `w:p`/`w:hyperlink`/`w:smartTag`. Folio
+    # preserves nested inline SDTs as `InlineSdt`, so misclassifying them
+    # as block here would produce a false `scope`/`childCount` divergence
+    # on every fixture with nested controls.
+    scope_parent = sdt_elem.getparent()
+    while scope_parent is not None and scope_parent.tag == SDT_CONTENT_TAG:
+        scope_parent = scope_parent.getparent()
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py]: Accepted and implemented. Treat both `w:sdtContent` and `w:sdt` as transparent when walking up to find the scope parent, so a nested inline SDT is classified by the surrounding `w:p`/`w:hyperlink`/`w:smartTag` instead of stopping at the outer `w:sdt`. Landed in c0efcfa15. CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/projection.ts [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+      if (child.type === "paragraph") {
+        visitParagraph(child);
+      } else {
+        visitTable(child);
+      }
+    }
+  };
+
+  const visitInlineSdt = (sdt: InlineSdt): void => {
+    sdts.push({
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts]: Accepted and implemented. Coalesce consecutive inline SDT segments that share a `SdtProperties` reference into one projection entry (summing `childCount`), so a wire `w:sdt` split by `pushInlineSdtSegments` no longer fans out into multiple SDT records. Landed in d5ff87859. CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/python_docx_project.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+
+# Parent tags that mark an SDT as inline-scoped. `w:sdtContent` and the
+# enclosing `w:sdt` are transparent for scope classification: when an inline
+# SDT is nested inside another inline SDT the parent chain is
+# `inner sdt → outer sdtContent → outer sdt → outer sdtContent → w:p`, so we
+# walk past both `w:sdtContent` and `w:sdt` ancestors to find the surrounding
+# paragraph/hyperlink/smartTag that decides scope. Folio preserves nested
+# inline SDTs as `InlineSdt`, so missing this step would misclassify them as
+# `block` and produce false `scope`/`childCount` divergences.
+INLINE_PARENT_TAGS = frozenset({qn("w:p"), qn("w:hyperlink"), qn("w:smartTag")})
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/python_docx_project.py:90]: Accepted and implemented. Added `w:ins`, `w:del`, `w:moveFrom`, and `w:moveTo` to `SCOPE_TRANSPARENT_TAGS` so the scope walker climbs past tracked-change wrappers and lands on the enclosing `w:p`/`w:hyperlink`/`w:smartTag` — matching folio's `parseParagraphContents`, which recurses through those wrappers and lifts non-run SDTs back to paragraph-level inline content. Landed in a3c791e0c.

CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/projection.ts [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+    for (const item of items) {
+      if (item.type !== "inlineSdt") {
+        flush();
+        continue;
+      }
+      if (pendingProps !== null && pendingProps !== item.properties) {
+        flush();
+      }
+      pendingProps = item.properties;
+      pendingChildCount += item.content.length;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts:168]: Accepted and implemented. Updated the python projector to collapse complex-field runs the same way folio does: every `w:r` from a `begin` fldChar through the matching `end` fldChar counts as one logical inline child, with a depth counter so nested complex fields (e.g. PAGEREF inside a TOC entry) don't close the outer span early. Verified end-to-end against a minimal inline-SDT-with-PAGE fixture (childCount: 1 on both sides). Landed in b8e979f90def5a2f96efe8c9ca167ca9d11a61d8. CC on behalf of @jan-kubica

File: packages/folio/scripts/differential/projection.ts [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+      pendingNested = [];
+      for (const child of nested) {
+        visitInlineSequence(child.content);
+      }
+    };
+
+    for (const item of items) {
+      if (item.type !== "inlineSdt") {
+        flush();
+        continue;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/folio/scripts/differential/projection.ts:162]: Accepted and implemented. Skipped the unconditional flush on non-inlineSdt siblings so split segments of the same wire `w:sdt` (sharing the same `SdtProperties` reference across lifted bookmark/comment/tracked-change markers) coalesce into one projection — matching python-docx's one entry per `w:sdt`. Different `SdtProperties` references still flush, and the trailing flush still runs at end of sequence. Landed in 97819c0d931b34804f7ddfce81a55ddc6990133f. CC on behalf of @jan-kubica
